### PR TITLE
FTP - Add proxy support to FTP, SFTP, and FTPS connectors

### PR DESCRIPTION
**Pull request by mjp426**
## Purpose

Adds proxy support to FTP, SFTP, and FTPS connectors

## References

References #600

## Changes

 * Add `withProxy` method on all three FTPSettings builders that takes a `java.net.Proxy` object
 * Pass the proxy object to the underlying client before the connection is made
 * Added tests for the three supported protocols via a proxy
 * Added squid proxy and configuration to docker-compose so this can be tested. 

## Background Context

The approach is relatively straight forward. The extra option fits neatly into to existing builder pattern and passing the proxy to the client at the start of the connect operation is the natural place for it to go. 

I've chosen to use a squid proxy in the testing as this is a widely used proxy.

### CODE REVIEW

**Comment by mjp426** [comment]
The build is now failing on the MiMa check. Could someone from the core team confirm I'm ok to add an exception for this change?

**Comment by mjp426** [comment]
Also, the unidoc task was failing because is was picking up an old version of commons-net from the hbase dependency. I've fixed it by upgrading that dependency, but perhaps there's a better way to do this?

**Comment by ennru** [comment]
Hi @mjp426 	- Welcome to Alpakka!

Thank you for working on this, as you've seen this feature has been requested for a long while.

The Mima exclusions for the settings classes are OK to add.
The Unidoc problem should be fixed in here instead:
https://github.com/akka/alpakka/blob/e48dd81f59fa00bd5beffabcc60720c83332c3c4/build.sbt#L69-L77

**Comment by mjp426** [comment]
Hi @ennru, thanks for the pointers. I've made the changes you suggested. 

My FTP tests are passing, but there are a few failing modules. I can't reproduce the failures locally, and they seem unrelated to my changes. Do you have any tips to get the tests green?

**Comment by mjp426** [comment]
Also, I think the dependency-change flag can be removed now.

File: ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala [hunk 1]
```diff
@@ -95,18 +99,20 @@ final class FtpSettings private (
...
     binary = binary,
     passiveMode = passiveMode,
-    configureConnection = configureConnection
+    configureConnection = configureConnection,
+    proxy = proxy
   )
 
   override def toString =
-    s"""FtpSettings(host=$host,port=$port,credentials=$credentials,binary=$binary,passiveMode=$passiveMode,configureConnection=$configureConnection)"""
+    s"""FtpSettings(host=$host,port=$port,credentials=$credentials,binary=$binary,passiveMode=$passiveMode,configureConnection=$configureConnection),proxy=$proxy"""
```
Comment (ennru) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala]: The `)` should move to the end. You may make this a single-" string and split it with one field per line, so this doesn't happen again.

File: ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala [hunk 2]
```diff
@@ -182,18 +192,20 @@ final class FtpsSettings private (
       credentials: FtpCredentials = credentials,
       binary: Boolean = binary,
       passiveMode: Boolean = passiveMode,
-      configureConnection: FTPSClient => Unit = configureConnection
+      configureConnection: FTPSClient => Unit = configureConnection,
+      proxy: Option[Proxy] = None
```
Comment (ennru) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala]:
```suggestion
      proxy: Option[Proxy] = proxy
```

File: ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala [hunk 3]
```diff
@@ -182,18 +192,20 @@ final class FtpsSettings private (
...
     binary = binary,
     passiveMode = passiveMode,
-    configureConnection = configureConnection
+    configureConnection = configureConnection,
+    proxy = proxy
   )
 
   override def toString =
-    s"""FtpsSettings(host=$host,port=$port,credentials=$credentials,binary=$binary,passiveMode=$passiveMode,configureConnection=$configureConnection)"""
+    s"""FtpsSettings(host=$host,port=$port,credentials=$credentials,binary=$binary,passiveMode=$passiveMode,configureConnection=$configureConnection),proxy=$proxy"""
```
Comment (ennru) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala]: Same thing with the `)`.

File: ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala [hunk 4]
```diff
@@ -250,25 +265,28 @@ final class SftpSettings private (
...
 
   private def copy(
       host: java.net.InetAddress = host,
       port: Int = port,
       credentials: FtpCredentials = credentials,
       strictHostKeyChecking: Boolean = strictHostKeyChecking,
       knownHosts: Option[String] = knownHosts,
-      sftpIdentity: Option[SftpIdentity] = sftpIdentity
+      sftpIdentity: Option[SftpIdentity] = sftpIdentity,
+      proxy: Option[Proxy] = None
```
Comment (ennru) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala]:
```suggestion
      proxy: Option[Proxy] = proxy
```

File: ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala [hunk 5]
```diff
@@ -95,18 +99,20 @@ final class FtpSettings private (
       credentials: FtpCredentials = credentials,
       binary: Boolean = binary,
       passiveMode: Boolean = passiveMode,
-      configureConnection: FTPClient => Unit = configureConnection
+      configureConnection: FTPClient => Unit = configureConnection,
+      proxy: Option[Proxy] = None
```
Comment (ennru) [inline ftp/src/main/scala/akka/stream/alpakka/ftp/model.scala]:
```suggestion
      proxy: Option[Proxy] = proxy
```

File: docs/src/main/paradox/ftp.md [hunk 6]
```diff
@@ -37,6 +37,8 @@ The example demonstrates optional use of `configureConnection` option available
 
 For non-anonymous connection, please provide an instance of @scaladoc[NonAnonFtpCredentials](akka.stream.alpakka.ftp.FtpCredentials$$NonAnonFtpCredentials) instead.
 
+For connection via a proxy, please provide an instance of `java.net.Proxy`.
```
Comment (ennru) [inline docs/src/main/paradox/ftp.md]: Ah, there is this. Maybe refer to the `withProxy` method.

**Comment by mjp426** [comment]
Good catch on the copy method. I've committed your suggestions.

**Comment by ennru** [comment]
Failures were 
https://github.com/akka/alpakka/issues/1909
https://github.com/akka/alpakka/issues/1801
https://github.com/akka/alpakka/issues/1723

**Comment by ennru** [comment]
Thank you! Great to see this long-standing feature request fixed.

**Pull request merged**